File: Nikhil/python/person_locator_agent.py
# person_locator_agent.py
from __future__ import annotations

import logging
from pydantic import BaseModel, Field

from toolable_react_agent import ToolableReactAgent
from tool import AbstractTool
from tool_call import ToolCall
from tool_call_result import ToolCallResult

# Logging configuration (mirrors project-wide settings)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger(__name__)


class PersonLocatorAgent(ToolableReactAgent):
    """ReAct agent able to locate a person’s city."""

    # ------------------------------------------------------------------ #
    # Nested tool
    # ------------------------------------------------------------------ #
    class LocatePersonTool(AbstractTool):
        """Tool that returns a hard-coded location for a person."""

        # ---------- parameters schema (Pydantic) ---------------------- #
        class Parameters(BaseModel):
            person: str = Field(
                ...,
                description="The name of the person you want to locate.",
            )
            thought: str = Field(
                ...,
                description="Your reasoning about why and how accomplish this step.",
            )

        # ---------- construction ------------------------------------- #
        def __init__(self) -> None:
            super().__init__(
                id_="locatePersonTool",
                description="Returns the city where a person is located.",
                parameters_cls=PersonLocatorAgent.LocatePersonTool.Parameters,
            )

        # ---------- invocation --------------------------------------- #
        def invoke(self, call: ToolCall) -> ToolCallResult:  # noqa: D401
            if not self.is_initialized():
                raise RuntimeError("Tool must be initialized.")

            person = self.get_string("person", call.arguments)
            return ToolCallResult.from_call(
                call,
                f"{person} is currently located in Padua, Italy.",
            )

    # ------------------------------------------------------------------ #
    # Agent initialisation
    # ------------------------------------------------------------------ #
    def __init__(self) -> None:
        super().__init__(
            id_=self.__class__.__name__,
            description="This tool is able to find a person's location.",
            tools=[PersonLocatorAgent.LocatePersonTool()],
            check_last_step=False,
        )

File: Nikhil/java/src/main/java/com/infosys/small/examples/PersonLocatorAgent.java
package com.infosys.small.examples;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.infosys.small.core.AbstractTool;
import com.infosys.small.core.ToolCall;
import com.infosys.small.core.ToolCallResult;
import com.infosys.small.react.ToolableReactAgent;
import com.kjetland.jackson.jsonSchema.annotations.JsonSchemaDescription;

import lombok.NonNull;

/**
 * A ReAct agent that can return location of a person.
 */
public class PersonLocatorAgent extends ToolableReactAgent {

	public static class LocatePersonTool extends AbstractTool {

		@JsonSchemaDescription("This is a class describing parameters for LocatePersonTool")
		public static class Parameters {

			@JsonProperty(required = true)
			@JsonPropertyDescription("The name of the person you want to locate.")
			public String person;

			@JsonProperty(required = true)
			@JsonPropertyDescription("Your reasoning about why and how accomplish this step.")
			public String thought;
		}

		public LocatePersonTool() {
			super("locatePersonTool", // Function name
					"Returns the city where a person is located.", // Function description
					LocatePersonTool.Parameters.class); // Function parameters
		}

		@Override
		public ToolCallResult invoke(@NonNull ToolCall call) throws Exception {

			// Tool implementation goes here.
			// In this example we simply return a random temperature.

			if (!isInitialized())
				throw new IllegalStateException("Tool must be initialized.");

			String person = getString("person", call.getArguments());
			return new ToolCallResult(call, person + " is currently located in Padua, Italy.");
		}
	} // GetCurrentWeatherTool class

	public PersonLocatorAgent() {
		super(PersonLocatorAgent.class.getSimpleName(), //
				"This tool is able to find temperature in a given town.", //
				List.of(new LocatePersonTool()), false);
	}
}